src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java

```diff
@@ -16,37 +16,65 @@
 
 import com.vmware.vipclient.i18n.VIPCfg;
 import com.vmware.vipclient.i18n.base.DataSourceEnum;
+import com.vmware.vipclient.i18n.base.cache.CacheItem;
 import com.vmware.vipclient.i18n.messages.api.opt.server.ComponentBasedOpt;
 import com.vmware.vipclient.i18n.messages.api.opt.server.StringBasedOpt;
 import com.vmware.vipclient.i18n.messages.dto.MessagesDTO;
 import com.vmware.vipclient.i18n.util.ConstantsKeys;
 import com.vmware.vipclient.i18n.util.JSONUtils;
 
 public class StringService {
-    private MessagesDTO dto    = null;
-
     Logger              logger = LoggerFactory.getLogger(StringService.class);
-
-    public StringService(MessagesDTO dto) {
-        this.dto = dto;
-    }
-
+    
     @SuppressWarnings("unchecked")
-    public String getString() {
-        String key = dto.getKey();
-        CacheService cacheservice = new CacheService(dto);
-        Map<String, String> map = cacheservice.getCacheOfComponent();
-        if (map == null) {
-            if (!cacheservice.isContainComponent()) {
-                Object o = new ComponentService(dto).getMessages();
-                map = (Map<String, String>) o;
-                cacheservice.addCacheOfComponent(map);
-            }
-        }
-        return (map == null || map.get(key) == null ? "" : map.get(key));
+    public String getString(MessagesDTO dto) {
+    	String key = dto.getKey();
+    	CacheService cacheService = new CacheService(dto);
+    	Map<String, String> cacheOfComponent = null;
+    	if (cacheService.isContainComponent()) { // Item is in cache
+    		CacheItem cacheItem = cacheService.getCacheOfComponent();
+    		cacheOfComponent = cacheItem.getCachedData();
+    		if (cacheService.isExpired()) { // cacheItem has expired
+    			// Update the cache in a separate thread
+    			populateCacheTask(cacheItem.getCacheProperties(), cacheService, dto); 		
+    		}
+    	} else { // Item is not in cache
+    		// Create a new HashMap to store cache properties.
+    		cacheOfComponent = populateCache(new HashMap<String, Object>(), cacheService, dto);
+       } 
+       return (cacheOfComponent == null || cacheOfComponent.get(key) == null ? "" : cacheOfComponent.get(key));
+    }
+    
+	private void populateCacheTask(Map<String, Object> cacheProps, final CacheService cacheService, MessagesDTO dto) {
+		Runnable task = () -> {
+			boolean running = true;
+	    	while (running) {
```

Comment:
'While' loop runs only once because running becomes false after the first round. Do you think a normal function call can do the task?

Reply:
No. The task is done in parallel in a separate thread so that the getString method happening in the main thread returns without waiting for the task to finish. This is by design. 

The while loop just makes sure uses the "running=false" in the finally block so that if any runtime exception happens,. the thread is not left alive.. which may lead to memory leak.